config/configparser/parser.go

```diff
@@ -122,60 +144,61 @@ func (l *Parser) Sub(key string) (*Parser, error) {
 }
 
-// deepSearch scans deep maps, following the key indexes listed in the
-// sequence "path".
-// The last value is expected to be another map, and is returned.
-//
-// In case intermediate keys do not exist, or map to a non-map value,
-// a new map is created and inserted, and the search continues from there:
-// the initial map "m" may be modified!
-// This function comes from Viper code https://github.com/spf13/viper/blob/5253694/util.go#L201-L230
-// It is used here because of https://github.com/spf13/viper/issues/819
-func deepSearch(m map[string]interface{}, path []string) map[string]interface{} {
-	for _, k := range path {
-		m2, ok := m[k]
-		if !ok {
-			// Intermediate key does not exist:
-			// create it and continue from there.
-			m3 := make(map[string]interface{})
-			m[k] = m3
-			m = m3
-			continue
-		}
-		m3, ok := m2.(map[string]interface{})
-		if !ok {
-			// Intermediate key is a value:
-			// replace with a new map.
-			m3 = make(map[string]interface{})
-			m[k] = m3
-		}
-		// continue search from here
-		m = m3
+// ToStringMap creates a map[string]interface{} from a Parser.
+func (l *Parser) ToStringMap() map[string]interface{} {
+	return l.k.Raw()
+}
+
+// decoderConfig returns a default mapstructure.DecoderConfig capable of parsing time.Duration
+// and weakly converting config field values to primitive types.  It also ensures that maps
+// whose values are nil pointer structs resolved to the zero value of the target struct (see
+// expandNilStructPointers). A decoder created from this mapstructure.DecoderConfig will decode
+// its contents to the result argument.
+func decoderConfig(result interface{}) *mapstructure.DecoderConfig {
+	return &mapstructure.DecoderConfig{
+		Result:           result,
+		Metadata:         nil,
+		TagName:          "mapstructure",
+		WeaklyTypedInput: true,
+		DecodeHook: mapstructure.ComposeDecodeHookFunc(
+			expandNilStructPointers(),
+			mapstructure.StringToTimeDurationHookFunc(),
+			mapstructure.StringToSliceHookFunc(","),
+		),
 	}
-	return m
 }
 
-// ToStringMap creates a map[string]interface{} from a Parser.
-func (l *Parser) ToStringMap() map[string]interface{} {
-	// This is equivalent to l.v.AllSettings() but it maps nil values.
-	// We can't use AllSettings here because of https://github.com/spf13/viper/issues/819
-
-	m := map[string]interface{}{}
-	// Start from the list of keys, and construct the map one value at a time.
-	for _, k := range l.v.AllKeys() {
-		value := l.v.Get(k)
-		path := strings.Split(k, KeyDelimiter)
-		lastKey := strings.ToLower(path[len(path)-1])
-		deepestMap := deepSearch(m, path[0:len(path)-1])
-		// Set innermost value.
-		deepestMap[lastKey] = value
+// In cases where a config has a mapping of something to struct pointers
+// we want nil values to resolve to the zero value of the underlying struct
+// just as we want nil values of a mapping of something to structs to resolve
+// to the zero value of that struct.
```

Comment:
I am not sure I understand the comment. Show an example perhaps?

Reply:
Clarified a bit more and added an example.  Test cases like these processor config tests exercise the decode hook: https://github.com/open-telemetry/opentelemetry-collector/blob/df76aa36cd124ba4ae1b3d1ac1941155f86ea568/internal/processor/filterset/regexp/config_test.go#L32